File: 02_LearnDSP-MATLAB/Sec04_PracticalMethods/case01_RemoveTrend.m
% Remove polynomial trend in signal
% Polynominal curve fitting and cancelling

% XiaoCY 2021-02-14

%%
clear;clc

fs = 100;
fsig = 1;
p0 = [-0.05, 1, -0.3];
t = (0:1/fs:10)';
trend = polyval(p0,t);
x0 = sin(2*pi*fsig*t)+trend;

%%
p = polyfit(t,x0,2);
trendfit = polyval(p,t);
x1 = x0-trendfit;
x2 = detrend(x0,2);

figure
plot(t,x0,'DisplayName','origin')
hold on
plot(t,x1,'DisplayName','polydetrend')
plot(t,x2,'--','DisplayName','detrend')
plot(t,trendfit,'DisplayName','trendfit')
plot(t,trend,'--','DisplayName','trend')
grid on
legend
xlabel('Time (s)')
ylabel('Signal')

File: 03_LearnDSP-Python/Sec04_PracticalMethods/case01_RemoveTrend.py
'''
Remove polynomial trend in signal
Polynominal curve fitting and cancelling

XiaoCY 2021-02-14
'''

#%%
import numpy as np
import matplotlib.pyplot as plt
import scipy.signal as sig

pi = np.pi

fs = 100.
fsig = 1.
p0 = np.array([-0.05, 1.0, -0.3])
t = np.arange(0,10,1/fs)
trend = np.polyval(p0,t)
x0 = np.sin(2*pi*fsig*t)+trend

#%%
p = np.polyfit(t,x0,2)
trendfit = np.polyval(p,t)
x1 = x0-trendfit

x2 = sig.detrend(x0)                # only support to remove linear trend

plt.figure()
plt.plot(t,x0,label='origin')
plt.plot(t,x1,label='polydetrend')
plt.plot(t,x2,'--',label='detrend')
plt.plot(t,trendfit,label='trendfit')
plt.plot(t,trend,'--',label='trend')
plt.grid()
plt.legend()
plt.xlabel('Time (s)')
plt.ylabel('Signal')
plt.show()